mozilla/blurts-server
# MNTOR-1878: API for location autocomplete

# References: 
Jira: [MNTOR-1878](https://mozilla-hub.atlassian.net/browse/MNTOR-1878)
Figma: [Enter details inputs](https://www.figma.com/file/CaEKIhvSJqf6KNIMzSkt40/Concepts-for-Monitor-MVP-Redesign?type=design&node-id=1205-61900&mode=design&t=xa4nmpd5bdTjhiM7-0)




# Description

This PR adds the data and API needed for the location auto-completion input during the “Personal info” step for the breaches scan.

**Location data**
In order to create a list of US cities with the features we need we can use the script `createLocationAutocompleteData` that downloads and parses data from [GeoNames](https://www.geonames.org/). In order to get a dataset that supports our use case we compile a JSON file from three exported table dumps: https://www.geonames.org/export/dump/.
- Locations: https://download.geonames.org/export/dump/US.zip
All US location features that we filter to only get populated places.
- Alternate names: https://download.geonames.org/export/dump/alternateNames.zip
Alternate location names that we can use to make searching for locations more user-friendly.
- Hierarchy: https://download.geonames.org/export/dump/hierarchy.zip
Used to filter out locations that have populated places as a parent and thus are not cities but districts.

**Location autocomplete API**
The API `location-autocomplete` takes a search query and provides location suggestions. With https://github.com/leeoniya/uFuzzy we search through the list of cities that we created with the script `createLocationAutocompleteData`.


# Screenshot

https://github.com/mozilla/blurts-server/assets/13835474/343fba64-443d-4af1-9ffd-fd6a7872b297

# How to test

1. Before running the site install the newly added dependencies.
```
npm install
```

2. Create location data: Running the script takes a few minutes. While running a temporary directory `tmp-data` is created and deleted after the script ran successfully. The resulting `locationAutocompleteData.json` file will be put in the root directory.
```
npx ts-node-esm src/scripts/build/createLocationAutocompleteData/index.ts
```

3. You can test the location autocompletion API using the test input on http://localhost:6060/admin/location-search. Make sure you are signed in and added yourself as an admin in the `.env` file under `ADMINS=`.
